add model initializer

rakelkar opened the pull request:
<!--  Thanks for sending a pull request!  Here are some tips for you:
1. If this is your first time, read our contributor guidelines https://www.kubeflow.org/docs/about/contributing/ and developer guide https://github.com/kubeflow/kfserving/blob/master/docs/DEVELOPER_GUIDE.md
2. If you want *faster* PR reviews, read how: https://git.k8s.io/community/contributors/guide/pull-requests.md#best-practices-for-faster-reviews
3. Follow the instructions for writing a release note: https://git.k8s.io/community/contributors/guide/release-notes.md
4. If the PR is unfinished, see how to mark it: https://git.k8s.io/community/contributors/guide/pull-requests.md#marking-unfinished-pull-requests
-->

**What this PR does / why we need it**:
YAAWH to inject a downloader for #148 

In an ideal world this will go away once we can convince knative on https://github.com/knative/serving/issues/4307

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #148 

**Special notes for your reviewer**:

1. Please confirm that if this PR changes any image versions, then that's the sole change this PR makes.

**Release note**:
<!--  Write your release note:
1. Enter your extended release note in the below block. If the PR requires additional action from users switching to the new release, include the string "action required".
2. If no release note is required, just write "NONE".
-->
```release-note

```

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/kubeflow/kfserving/156)
<!-- Reviewable:end -->

rakelkar commented:
@ellis-bigelow i have download working for gcs but the TF sample is not working for me :( it fails with this error:
```
kubectl logs flowers-sample-default-6l8d7-deployment-ffdf7fc58-z5jbp -c user-container
2019-06-17 07:09:21.110764: I tensorflow_serving/model_servers/server.cc:82] Building single TensorFlow model file config:  model_nam
e: flowers-sample model_base_path: /mnt/model
2019-06-17 07:09:21.110929: I tensorflow_serving/model_servers/server_core.cc:461] Adding/updating models.
2019-06-17 07:09:21.110951: I tensorflow_serving/model_servers/server_core.cc:558]  (Re-)adding model: flowers-sample
2019-06-17 07:09:21.111215: W tensorflow_serving/sources/storage_path/file_system_storage_path_source.cc:268] No versions of servable
 flowers-sample found under base path /mnt/model
```
This seems to be a well known error if the model was not exported correctly e.g. https://github.com/tensorflow/models/issues/3530

So I was wondering what you have uploaded to gs://kfserving-samples/models/tensorflow/flowers? Did you freeze the model with tag set to serve? And how come it works if when I point the server straight at the bucket path - thats odd??

Here is what the downloader INIT container did (do the contents look accurate to you?):
```
$ kubectl logs flowers-sample-default-6l8d7-deployment-ffdf7fc58-z5jbp -c downloader
INFO:root:Copying contents of gs://kfserving-samples/models/tensorflow/flowers to local
Downloading from: gs://kfserving-samples/models/tensorflow/flowers  to /mnt/model
/mnt/model/0001/saved_model.pb
/mnt/model/0001/variables/variables.data-00000-of-00001
/mnt/model/0001/variables/variables.index
INFO:root:Successfully copied gs://kfserving-samples/models/tensorflow/flowers to /mnt/model
```

rakelkar commented on pkg/webhook/admission/deployment/downloader_injector.go:
```diff
@@ -0,0 +1,105 @@
+	defaultMountName         = "kfserving-download-location"
+	defaultMountPath         = "/mnt"
+	userContainerName        = "user-container"
+	downloadContainerImage   = "kcorer/kfdownloader"
```
Need to figure out how to fix this...? maybe put it into a config-map and then somehow do some ko like thing for the dev loop?

hougangliu replied:
agree, I think in kfserving release  we should add a kfserving configmap about configuration such as downloadContainerImage image, TensorflowServingImageName +DefaultTensorflowRuntimeVersion and so on, and in manager deployment adds related env whose value refers the configmap counterpart.
So I suggest `buildInitContainer` should try to get `downloadContainerImage` by ENV, if doesn't exist, use `kcorer/kfdownloader:latest` then.
Thus, in each release, we only need update the released configmap entry without updating our source code.

yuzisun replied:
we already have ConfigMap, we can add the download image there.

rakelkar replied:
Will do as a separate commit, tracked by #189

rakelkar commented on pkg/webhook/admission/deployment/accelerator_injector.go:
```diff
@@ -16,61 +16,19 @@ limitations under the License.
 )
 
-// Handle decodes the incoming Pod and executes mutation logic.
-func (mutator *Mutator) Handle(ctx context.Context, req types.Request) types.Response {
```
I created a single mutator that calls both injectors instead of creating two separate webhooks... @ellis-bigelow  is this the right pattern? how can we get good logs for this so its easy to debug issues with the webhook - to figure out what has run, what hasnt, and who did what...?

ellis-bigelow replied:
I was thinking about this. I think it's acceptable otherwise we will be making multiple http calls in a row to do mutation.

hougangliu commented on pkg/webhook/admission/deployment/deployment_mutator.go:
```diff
@@ -0,0 +1,67 @@
+
+// Handle decodes the incoming Pod and executes mutation logic.
+func (mutator *Mutator) Handle(ctx context.Context, req types.Request) types.Response {
+	deployment := &appsv1.Deployment{}
```
We should validate ModelURI in validation webhook, for example, now we only support s3, gs and local path, the ModelURI must match the three storage format

rakelkar replied:
Great suggestion - done!

rakelkar commented on pkg/webhook/admission/deployment/model_provisioning_injector_test.go:
```diff
@@ -0,0 +1,265 @@
+				},
+			},
+		},
+		"ProvisionerInjectedAndMountsPvc": {
```
@hougangliu  check it out - pvc support implemented by adding the mounts and converting src path to be relative to the pvc mount..

hougangliu commented on pkg/webhook/admission/deployment/model_provisioning_injector.go:
```diff
@@ -0,0 +1,171 @@
+	}
+
+	if userContainerIndex < 0 {
+		return fmt.Errorf("aargh no conuser container: %s", srcURI)
```
in this negative case, we can ignore it, taking it as the non-knative-generated deployment. Maybe a warning message is enough

rakelkar replied:
The annotation is private - therefore if it has reached this point it is an error

hougangliu commented on pkg/webhook/admission/deployment/model_provisioning_injector.go:
```diff
@@ -0,0 +1,171 @@
+	}
+
+	// Only inject provisioning for supported URIs
+	if !strings.HasPrefix(srcURI, "gs://") &&
```
I suggest we validate kfservice ModelUri in validating webhook instead of here

rakelkar replied:
done!

ellis-bigelow commented on python/commands/download:
```diff
@@ -0,0 +1,13 @@
+#!/usr/bin/env python3
```
Can we put this in python/predictor-initializer?

rakelkar replied:
put into model-initializer.. as we discussed that makes more sense for now

rakelkar commented on python/commands/download:
```diff
@@ -0,0 +1,13 @@
+src_uri = sys.argv[1]
+dest_path = sys.argv[2]
+
+print("Downloading from:", src_uri, " to", dest_path)
```
the formatting is actually correct since the x,y gets a space for free... :)  however will update to use logging

ellis-bigelow commented on pkg/webhook/admission/deployment/deployment_mutator.go:
```diff
@@ -0,0 +1,67 @@
+	return third_party.PatchResponseFromRaw(req.AdmissionRequest.Object.Raw, patch)
+}
+
+func mutate(deployment *appsv1.Deployment) error {
```
Can we make a slice of mutators and loop over them? :)

rakelkar replied:
done :)

ellis-bigelow commented on pkg/webhook/admission/deployment/deployment_mutator.go:
```diff
@@ -0,0 +1,67 @@
+/*
```
Nit: call this call mutator.go

rakelkar replied:
done!

ellis-bigelow commented on pkg/webhook/admission/deployment/model_provisioning_injector.go:
```diff
@@ -0,0 +1,171 @@
+const (
+	ModelProvisioningContainerName    = "model-provisioner"
+	ModelProvisioningVolumeName       = "kfserving-provision-location"
+	ModelProvisioningContainerImage   = "kcorer/kfdownloader"
```
Let's make this configurable & a standardized link. Just cut me an issue to setup to automated build.

rakelkar replied:
made it standard... trying to figure out how to make it configurable.. the webhook doesn't load config right now.. can point me to an example of how to do this correctly?

ellis-bigelow commented on pkg/webhook/admission/deployment/model_provisioning_injector.go:
```diff
@@ -0,0 +1,171 @@
+	annotations := deployment.Spec.Template.ObjectMeta.Annotations
+	podSpec := &deployment.Spec.Template.Spec
+
+	// Only inject if the required annotations are set
```
This is pretty awkward. Annotations are for enabling non-default behavior yet isn't this happening by default?

rakelkar replied:
As discussed this is an internal annotation

ellis-bigelow commented on pkg/webhook/admission/deployment/model_provisioning_injector.go:
```diff
@@ -0,0 +1,171 @@
+		mountPath = constants.DefaultModelLocalMountPath
+	}
+
+	// Only inject provisioning for supported URIs
```
I thought we were going to run the injector in all cases? If it isn't a supported URI, our validating admission controller should reject it. Otherwise we're going to need to keep editing this code every time we add new storage support.

rakelkar replied:
Moved validation to the validator

ellis-bigelow commented on pkg/constants/constants.go:
```diff
@@ -40,7 +40,9 @@ var (
-	KFServiceGKEAcceleratorAnnotationKey = KFServingAPIGroupName + "/gke-accelerator"
+	KFServiceGKEAcceleratorAnnotationKey             = KFServingAPIGroupName + "/gke-accelerator"
+	KFServiceModelProvisioningSourceURIAnnotationKey = KFServingAPIGroupName + "/provisioning/sourceURI"
+	KFServiceModelProvisioningMountPathAnnotationKey = KFServingAPIGroupName + "/provisioning/mountPath"
```
This is an odd pattern for annotations. We should probably chat offline about the annotation and defaulting behavior.

rakelkar replied:
Have made it private as discussed

ellis-bigelow commented on pkg/apis/serving/v1alpha1/framework.go:
```diff
@@ -41,6 +42,10 @@ var (
 	DefaultCPURequests    = resource.MustParse("1")
 )
 
+func (m *ModelSpec) MapSourceUri() (sourceURI string, localPath string, ok bool) {
```
Would love to see this collapsed as GetModelUri and avoid the extra config

rakelkar replied:
collapsed to GetModelUri

ellis-bigelow commented on pkg/constants/constants.go:
```diff
@@ -40,7 +40,9 @@ var (
 var (
-	KFServiceGKEAcceleratorAnnotationKey = KFServingAPIGroupName + "/gke-accelerator"
+	KFServiceGKEAcceleratorAnnotationKey             = KFServingAPIGroupName + "/gke-accelerator"
+	KFServiceModelProvisioningSourceURIAnnotationKey = KFServingAPIGroupName + "/provisioning/sourceURI"
```
nit: uri

rakelkar replied:
lint wanted it to be caps

ellis-bigelow commented on pkg/constants/constants.go:
```diff
@@ -40,7 +40,9 @@ var (
-	KFServiceGKEAcceleratorAnnotationKey = KFServingAPIGroupName + "/gke-accelerator"
+	KFServiceGKEAcceleratorAnnotationKey             = KFServingAPIGroupName + "/gke-accelerator"
+	KFServiceModelProvisioningSourceURIAnnotationKey = KFServingAPIGroupName + "/provisioning/sourceURI"
+	KFServiceModelProvisioningMountPathAnnotationKey = KFServingAPIGroupName + "/provisioning/mountPath"
```
Let's drop this annotation for now.

rakelkar replied:
dropped

rakelkar commented on python/commands/download:
```diff
@@ -0,0 +1,13 @@
+dest_path = sys.argv[2]
+
+print("Downloading from:", src_uri, " to", dest_path)
+kfserving.Storage.download(src_uri, dest_path)
```
changed to use logging

rakelkar commented:
@ellis-bigelow resolved all comments but one (need to make container configurable):
Removed Map method with GetModelSourceURI
Removed second annotation and made path a const
Pointing to gcs container (note: will need you to push this ASAP after commit, else all are broken) #183 
Removed makefile
Renamed to model-initializer

ellis-bigelow commented:
/hold
until my autobuild PR is in

ellis-bigelow commented on pkg/apis/serving/v1alpha1/kfservice_validation.go:
```diff
@@ -29,6 +31,11 @@ const (
+)
+
+var (
+	SuppotedModelSourceURIPrefixList = []string{"gs://", "s3://", "pvc://", "file://"}
```
What is the file:// prefix used for?

rakelkar replied:
see storage.py
```
_LOCAL_PREFIX = "file://"

...

        if uri.startswith(_LOCAL_PREFIX) or os.path.exists(uri):
            return Storage._download_local(uri)
```

ellis-bigelow commented on pkg/apis/serving/v1alpha1/kfservice_validation.go:
```diff
@@ -77,12 +84,35 @@ func validateModelSpec(spec *ModelSpec) error {
+
+	// local path (not some protocol?)
+	if !regexp.MustCompile("\\w+?://").MatchString(modelSourceURI) {
+		return nil
```
We gotta error if this fails, right?

ellis-bigelow replied:
"Unknown model source"?

rakelkar replied:
so if MustCompile fails - it panics... if the regex match fails it means it is not a protocol.. so am assuming that means its some sort of local path :-)

ellis-bigelow commented on pkg/apis/serving/v1alpha1/kfservice_validation.go:
```diff
@@ -29,6 +31,11 @@ const (
+)
+
+var (
+	SuppotedModelSourceURIPrefixList = []string{"gs://", "s3://", "pvc://", "file://"}
```
Suppoted -> Supported

ellis-bigelow commented on pkg/constants/constants.go:
```diff
@@ -43,6 +43,12 @@ var (
+// KFService Internal Annotations
+var (
+	KFServicePrivateAnnotationsPrefix                       = KFServingAPIGroupName + "/internal"
+	KFServiceModelInitializerSourceURIInternalAnnotationKey = KFServicePrivateAnnotationsPrefix + "/model-initializer/sourceURI"
```
This is a variable name given by the gods.

rakelkar replied:
LOL!... suggestions welcome :)

ellis-bigelow commented on pkg/constants/constants.go:
```diff
@@ -52,22 +58,25 @@ var (
 )
 
+// constants.DefaultModelLocalMountPath is where models will be mounted by default unless overriden by a framework
+const DefaultModelLocalMountPath = "/mnt/kfserving-model"
```
How about /mnt/models?

ellis-bigelow commented on python/kfserving/kfserving/storage.py:
```diff
@@ -27,23 +27,25 @@
     @staticmethod
-    def download(uri: str) -> str:
-        logging.info("Copying contents of %s to local", uri)
+    def download(uri: str, out_dir: str = None) -> str:
```
ooo can you mkdtemp inline ;)

rakelkar replied:
would be a terrible idea... because python...
>Python’s default arguments are evaluated once when the function is defined, not each time the function is called (like it is in say, Ruby). This means that if you use a mutable default argument and mutate it, you will and have mutated that object for all future calls to the function as well.

https://docs.python-guide.org/writing/gotchas/

ellis-bigelow replied:
This actually sounds good to me. The outdir is scoped to the lifetime of the program :)

ellis-bigelow commented:
More or less ready to commit. Let me work with you tomorrow to get an autobuild working before we push this. It's fine if the build is failing until this releases, we just need to make sure that it's working

rakelkar commented:
> we just need to make sure that it's working

Tested TensorRT and Tensorflow samples and they work...

rakelkar commented:
/retest

yuzisun commented on pkg/apis/serving/v1alpha1/framework_tensorflow.go:
```diff
@@ -58,7 +63,7 @@ func (t *TensorflowSpec) CreateModelServingContainer(modelName string, config *F
 			"--rest_api_port=" + TensorflowServingRestPort,
 			"--model_name=" + modelName,
-			"--model_base_path=" + t.ModelURI,
+			"--model_base_path=" + constants.DefaultModelLocalMountPath,
```
Does this force tensorflow serving to use model initializer ?

rakelkar replied:
It does yes.

rakelkar commented:
/retest

ellis-bigelow commented:
Rakesh, a couple thoughts.

1) Should the dockerfile go in model initalizer folder? I don't like this business of having weirdly named dockerfiles outside.
2) Can you build me an image I can copy to gcr
3) Can you add the build script here to the Makefile?

rakelkar commented:
@ellis-bigelow  cannot move the Dockerfile in since it depends on the kfserving folder and the build context doesnt include stuff from outside the folder path.

Container is here: https://hub.docker.com/r/kcorer/kfdownloader/

rakelkar commented:
@ellis-bigelow keep hold - I think this is may be causing us to hit a race in knative :-(
https://github.com/knative/serving/issues/4564

ellis-bigelow commented:
Any extra infra @rakelkar

rakelkar commented:
Race fixed by https://github.com/kubeflow/kfserving/pull/202

rakelkar commented on pkg/constants/constants.go:
```diff
@@ -43,6 +43,12 @@ var (
+// KFService Internal Annotations
+var (
+	KFServicePrivateAnnotationsPrefix                       = "internal." + KFServingAPIGroupName
+	KFServiceModelInitializerSourceURIInternalAnnotationKey = KFServicePrivateAnnotationsPrefix + "/model-initializer-sourceURI"
```
updated, feel free to suggest if can think of something better... :)

rakelkar commented:
/retest

yuzisun commented on config/default/crds/serving_v1alpha1_kfservice.yaml:
```diff
@@ -68,14 +68,14 @@ spec:
                       type: string
                     resources:
-                      description: Defaults to requests and limits of 1CPU, 2Gb MEM.
+                      description: Defaults to latest Runtime Version.
```
Why the description changed here?

rakelkar replied:
For some reason codegen didnt run on #153? if you look at the code, this is what it should be:
```
	// Defaults to latest Runtime Version.
	Resources v1.ResourceRequirements `json:"resources,omitempty"`
```

yuzisun replied:
oh it is wrong comments then

yuzisun commented on config/default/crds/serving_v1alpha1_kfservice.yaml:
```diff
@@ -68,14 +68,14 @@ spec:
+                      description: Defaults to latest Runtime Version.
                       type: object
                     runtimeVersion:
+                      description: Defaults to  PyTorchModel
```
sam here

yuzisun commented on pkg/apis/serving/v1alpha1/framework_scikit.go:
```diff
@@ -43,7 +48,7 @@ func (s *SKLearnSpec) CreateModelServingContainer(modelName string, config *Fram
 		Args: []string{
 			"--model_name=" + modelName,
-			"--model_dir=" + s.ModelURI,
+			"--model_dir=" + constants.DefaultModelLocalMountPath,
```
I am wondering how this would work with knative revisions, if user only canary model uri but here we are mutating to a constant model path, does this mean knative would never create a new revision ?

rakelkar replied:
the annotation is different, that causes a new revision to be created...
```
$ kubectl get revision
NAME                                   SERVICE NAME                           GENERATION   READY   REASON
tensorrt-simple-string-default-6c6p9   tensorrt-simple-string-default-6c6p9   2            True    
tensorrt-simple-string-default-b5w49   tensorrt-simple-string-default-b5w49   1            True    
```

yuzisun replied:
Ah, can we leave some comments for that annotation ? It is quite difficult for people to figure out what's going on there.

rakelkar commented:
is good to merge @yuzisun @ellis-bigelow

ellis-bigelow commented:
/approve. 

@yuzisun, do you mind tossing the lgtm if you're ready?

rakelkar commented:
There is already a comment on
pkg/controller/kfservice/resources/knative/configuration.go
Have added additional comment in framework.go and updated the comment in
constants.go

On Sat, Jun 29, 2019 at 8:24 PM Dan Sun <notifications@github.com> wrote:

> *@yuzisun* commented on this pull request.
> ------------------------------
>
> In pkg/apis/serving/v1alpha1/framework_scikit.go
> <https://github.com/kubeflow/kfserving/pull/156#discussion_r298816734>:
>
> > @@ -43,7 +48,7 @@ func (s *SKLearnSpec) CreateModelServingContainer(modelName string, config *Fram
>  		Resources: s.Resources,
>  		Args: []string{
>  			"--model_name=" + modelName,
> -			"--model_dir=" + s.ModelURI,
> +			"--model_dir=" + constants.DefaultModelLocalMountPath,
>
> Ah, can we leave some comments for that annotation ? It is quite difficult
> for people to figure out what's going on there.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/kubeflow/kfserving/pull/156?email_source=notifications&email_token=AA2CTBMSKJUEEUUDZUOIBQ3P5ARM7A5CNFSM4HXJLWD2YY3PNVWWK3TUL52HS4DFWFIHK3DMKJSXC5LFON2FEZLWNFSXPKTDN5WW2ZLOORPWSZGOB5BR2LI#discussion_r298816734>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AA2CTBMBUZWYGOSMFI7BDFTP5ARM7ANCNFSM4HXJLWDQ>
> .
>


-- 
Rakesh Kelkar

rakelkar commented:
/retest

rakelkar commented:
/unhold

rakelkar commented:
@ellis-bigelow how do i get rid of the hold tag? I tried this already:
/unhold

rakelkar commented:
also doesn't look like your approve worked?

rakelkar commented:
/hold cancel

ellis-bigelow commented:
/approve

ellis-bigelow commented:
/lgtm
